### ISSUE TITLE
Python references that go out of scope might not release the memory allocated by _upb_Arena_SlowMalloc 

### ISSUE THREAD

**Issue by tvalentyn** [protocolbuffers/upb#1526]
Repro:

1. In a clean environment, run `pip install apache-beam==2.51.0`.
2. Download  pipeline.pb (see attached).
3. Create a `memleak_repro.py`

```
from apache_beam.portability.api import beam_runner_api_pb2
from apache_beam.io.gcp import gcsio
from apache_beam.io.gcp import gcsfilesystem

if len(sys.argv) <= 1:
  print("Specify proto path!")
  sys.exit(1)

path = sys.argv[1]
if path.startswith('gs'):
  open = gcsio.GcsIO().open

with open(path, 'rb') as f:
  pipeline = beam_runner_api_pb2.Pipeline()
  pipeline.ParseFromString(f.read())

for transform in pipeline.components.transforms:
  for _ in range(1000000):
    reference = pipeline.components.transforms[transform].outputs
```
Run ```python memleak_repro.py pipeline.pb```, observe RAM usage in top/htop, etc, increasing the number of iterations as necessary to keep the process running. 

5. For profiling, run: 

```pip install memray```

```memray run  -o output.bin --force memleak_repro.py pipeline.pb ;  memray table --leak output.bin -o table.html --force ; memray flamegraph --leak output.bin -o flamegraph.html -f```

7. Open table.html, double click on the Size column, note ~300MB reported leaked memory. 

Leak doesn't happen with  `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python` set.  

Leak can also be observed with C-stack tools, such as tcmalloc:

```
apt install google-pprof
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc.so.4:$LD_PRELOAD  HEAPPROFILE=`pwd`/  python memleak_repro.py pipeline.pb
google-pprof /home/valentyn/.pyenv/versions/py310/bin/python --inuse_space  _219720.0005.heap  --base=_219720.0004.heap 
Using local file /home/valentyn/.pyenv/versions/py310/bin/python.
Using local file _219720.0005.heap.
Welcome to pprof!  For help, type 'help'.
(pprof) top
Total: 1152.0 MB
  1152.0 100.0% 100.0%   1152.0 100.0% _upb_Arena_SlowMalloc
(pprof) 
```

Using memray in `--native` mode also points to  `_upb_Arena_SlowMalloc` (you might want to reduce the number of iterations to observe the leak.

**Comment by tvalentyn** [comment]
cc: @haberman @ericsalo

**Comment by tvalentyn** [comment]
[pipeline.pb.zip](https://github.com/protocolbuffers/upb/files/12898151/pipeline.pb.zip)

**Comment by tvalentyn** [comment]
This is the proto we use: https://github.com/apache/beam/blob/47d0fd566f86aaad35d26709c52ee555381823a4/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L91

I am sure the repro can be simplified further.

**Comment by anandolee** [comment]
@tvalentyn can you help create a repo that not depend on apache-beam? Thanks

**Comment by tvalentyn** [comment]
I am OOO atm and likely won't get to it before next week, but cc'ing: @anandinguva in case he has time and interested.

**Comment by tvalentyn** [comment]
Interesting. I gave it a try, and I was not able to recreate the leak in isolation while using essentially the same proto.

I'll check whether this has to do with how beam compiles proto stubs or something else.

**Comment by haberman** [comment]
> I'll check whether this has to do with how beam compiles proto stubs or something else.

That would be really helpful, thanks!

**Comment by tvalentyn** [comment]
Got a smaller repro. Appears that there some interaction between protobuf and dill. Beam uses old version of dill, but the issue is reproducible even with latest dill. Good news is there is some work underway for Beam to not depend on Dill and at least be able to use cloudpickle as default.. 

```
pip install dill protobuf grpcio-tools dill memray 
```

```
import dill # Leak doesn't happen without this import.
import sys
from google.protobuf import text_format

proto_def="""
syntax = "proto3";

message MessageWithAMap {
  map<string, string> some_map = 1;
}
"""

with open("repro.proto", "w") as f:
  f.write(proto_def)

from grpc_tools import protoc
args = [
    sys.executable, # unused
    "repro.proto",
    "--proto_path", ".", "--python_out", "."]
protoc.main(args)

import repro_pb2
from repro_pb2 import MessageWithAMap

proto="""
some_map {
  key: "Length of this key increases the leak"
  value:"No observable impact on the leak"
}
"""

message = MessageWithAMap()
text_format.Parse(proto, message)

key = list(message.some_map)[0]
for _ in range(500000):
  unused_values = message.some_map[key] # memory leaks here

```

**Comment by tvalentyn** [comment]
Leak doesn't happen with protobuf==3.20.3 or with `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python`.

**Comment by haberman** [comment]
I think I see the root cause. In map lookup, we use the message's arena to convert the Python `key` object to raw C data that we can use to perform the lookup in upb. But this data will never be freed until the message is freed: https://github.com/protocolbuffers/protobuf/blob/022d223488c0807e11d1c2509123901f5cc6b0b6/python/map.c#L203-L205

We should use a temporary arena to perform the conversion. Only at the point where we are allocating non-temporary data should we use the message's arena.

**Comment by tvalentyn** [comment]
That explains why `import __main__ as _main_module ` has an effect. Leak is still happening without this statement, but it makes it easier to register the leak with the profiler as references to `message` persist longer.

**Comment by tvalentyn** [comment]
Indeed, in `top` memory keeps growing without `import __main__ as _main_module` and memray also registers the leak when used without `--leak` flag.

**Comment by tvalentyn** [comment]
`PyUpb_MapContainer_Get`  also has this leak https://github.com/protocolbuffers/protobuf/blob/022d223488c0807e11d1c2509123901f5cc6b0b6/python/map.c#L259-L264

**Comment by tvalentyn** [comment]
And probably several other places in `map.c` where PyUpb_PyToUpb is used.
